Prune game tree by removing lines that include specific actions

bkushigian opened the pull request:
This PR addresses Issue #14.

bkushigian commented:
@b-inary I think this is correct. Let me know if tests are sufficient. I don't test for memory usage change since I didn't know if that was a stable feature yet and don't want to make brittle tests.

b-inary commented on src/game.rs:
```diff
@@ -966,6 +966,117 @@ impl PostFlopGame {
+    pub fn remove_lines(&mut self, lines: &Vec<Vec<Action>>) -> Result<(), String> {
+        if self.state <= State::Uninitialized {
+            return Err("Game is not successfully initialized".to_string());
+        } else if self.state == State::MemoryAllocated {
```
Should accept only `State::TreeBuilt`, not `State::Solved`
```suggestion
        } else if self.state >= State::MemoryAllocated {
```

bkushigian replied:
good catch, thanks!

b-inary commented on src/game.rs:
```diff
@@ -966,6 +966,117 @@ impl PostFlopGame {
+    /// Remove lines after building the PostFlopGame but before allocating memory.
+    /// This allows the removal of chance-specific lines (e.g., remove overbets on
+    /// board-pairing turns) which we cannot do while building an action tree.
+    pub fn remove_lines(&mut self, lines: &Vec<Vec<Action>>) -> Result<(), String> {
```
I prefer `&[Vec<Action>]` instead of `&Vec<Vec<Action>>`
```suggestion
    pub fn remove_lines(&mut self, lines: &[Vec<Action>]) -> Result<(), String> {
```

bkushigian replied:
:+1:

b-inary commented on src/game.rs:
```diff
@@ -966,6 +966,117 @@ impl PostFlopGame {
+        let index = search_result.unwrap();
+        if line.len() > 1 {
+            let result = self.remove_line_recursive(&mut node.children[index].lock(), &line[1..]);
+            if line.len() == 2 {}
```
Unnecessary line?

bkushigian replied:
yup, thanks :)

b-inary commented on src/game.rs:
```diff
@@ -966,6 +966,117 @@ impl PostFlopGame {
+        // STEP 1
+        let mut info = BuildTreeInfo {
+            memory_usage_nodes: mem::size_of::<PostFlopNode>() as u64,
+            num_storage_actions: 0,
```
Should be `self.num_private_hands(node.player as usize)` when the last action of the `line` is not a chance action.
```suggestion
            num_storage_actions: self.num_private_hands(node.player as usize),
```

b-inary commented on src/game.rs:
```diff
@@ -966,6 +966,117 @@ impl PostFlopGame {
+        }
+
+        Ok(info)
+        //Err("Not implemented".to_string())
```
Unnecessary line?

bkushigian replied:
yuuup!

b-inary commented on src/game.rs:
```diff
@@ -966,6 +966,117 @@ impl PostFlopGame {
+        node.children.shrink_to_fit();
+
+        // STEP 3
+        if node.is_chance() {
+            node.num_elements = match node.cfvalue_storage(PLAYER_OOP as usize) {
+                CfValueStorage::None => 0,
+                CfValueStorage::Sum => self.num_private_hands(PLAYER_OOP as usize),
+                CfValueStorage::All => {
+                    node.num_actions() * self.num_private_hands(PLAYER_OOP as usize)
+                }
+            };
```
Removing a chance action is somewhat misleading, so I do not want to support it. I prefer returning an error rather than processing it.

bkushigian replied:
Okay, I added an error check above. I agree that it shouldn't be supported

b-inary commented on src/game.rs:
```diff
@@ -3092,6 +3203,167 @@ mod tests {
+        game.play(game.node().actions.binary_search(&Action::Check).unwrap());
+        game.play(game.node().actions.binary_search(&Action::Check).unwrap());
+        game.play(2);
+        assert!(game.available_actions() == &[Action::Check]);
+        assert!(game.node().children.len() == 1);
```
I prefer `assert_eq!` instead of `assert!` when available (omit comments on other lines).

b-inary commented on src/game.rs:
```diff
@@ -3092,6 +3203,167 @@ mod tests {
+        ];
+
+        let res = game.remove_lines(&lines);
+        println!("result: {res:?}");
```
I prefer not to use the `println!` macro in normal tests.

bkushigian replied:
good catch, thought i got rid of all the printing

b-inary commented on src/game.rs:
```diff
@@ -3092,6 +3203,167 @@ mod tests {
+        assert!(game.available_actions() == &[Action::Check, Action::Bet(30)]);
+        assert!(game.node().children.len() == 2);
+
+        solve(&mut game, 10, 0.05, true);
```
I prefer not to print progress on normal tests (`solve_pio_preset_*` tests print progress because they are ignored tests).

bkushigian replied:
good point, done!

b-inary commented on src/game.rs:
```diff
@@ -3092,6 +3203,167 @@ mod tests {
+        assert!(game.available_actions() == &[Action::Check, Action::Bet(30)]);
+        assert!(game.node().children.len() == 2);
+
+        solve(&mut game, 10, 0.05, true);
```
Same as the L3259.

bkushigian replied:
:+1:

b-inary commented on src/game.rs:
```diff
@@ -3092,6 +3203,167 @@ mod tests {
         assert!(game.is_err());
     }
 
+    #[test]
+    fn remove_lines() {
```
Probably this test is unnecessary because it is covered by the `remove_lines2` test.

bkushigian replied:
gone!

b-inary commented on src/game.rs:
```diff
@@ -3092,6 +3203,167 @@ mod tests {
+            ..Default::default()
+        };
+
+        // Simple tree: force checks until river, where OOP can bet 1/2 pot
```
wrong comment

bkushigian replied:
fixed

b-inary reviewed (commented):
Thanks so much for your contribution! I have done some code reviews, so please check them out.

b-inary commented:
Also, this procedure can produce unexpected terminal nodes, that is, nodes without children nodes but the state is not terminal. The `remove_lines` method might have to return an error when the input action is the last one in the input node (`ActionTree` can add an action, but `PostFlopGame` cannot).

bkushigian commented:
Okay, addressed the above issues! I saw that clippy failed on my last push. I ran it locally and it was complaining about stuff I didn't touch. If it complains again we can look in more detail, but hopefully it will just work

bkushigian commented:
> Also, this procedure can produce unexpected terminal nodes, that is, nodes without children nodes but the state is not terminal. The remove_lines method might have to return an error when the input action is the last one in the input node (ActionTree can add an action, but PostFlopGame cannot).

good point, i need to address this still

bkushigian commented:
addressed

b-inary commented:
Merged. Thank you very much for your response!
